### PR TITLE
Proper error handling

### PR DESCRIPTION
**Pull request by ecomba**
This PR changes the way we did error handling.

Please note this PR is still a draft

### CODE REVIEW

**Review by onlyhavecans — CHANGES_REQUESTED** [review]
The error messages returned are inconsistent.

The hardcoded ones are fine but the ones that use variables are wrapping the error Enum names in extra quotes which will make this hard to use in code.
```NotFound("\"Domain `example.com` not found\"")```

When displayed they also have the extra `Message: ` as well as the quotes
```Message: "Domain `example.com` not found"```

With Transport Errors, it also doubles the error message. Giving the error message, then giving it again in parentheses

`Transport Error – https://api.sandbox.dnsimple.com/v2/whoami: Dns Failed: resolve dns name 'api.sandbox.dnsimple.com:443': failed to lookup address information: nodename nor servname provided, or not known(https://api.sandbox.dnsimple.com/v2/whoami: Dns Failed: resolve dns name 'api.sandbox.dnsimple.com:443': failed to lookup address information: nodename nor servname provided, or not known)`



I haven't tested all the possible errors yet but I have started expanding the API examples to help with intigrations tests.

**Comment by ecomba** [comment]
> The error messages returned are inconsistent.
> 
> The hardcoded ones are fine but the ones that use variables are wrapping the error Enum names in extra quotes which will make this hard to use in code. `` NotFound("\"Domain `example.com` not found\"") ``
> 
> When displayed they also have the extra `Message: ` as well as the quotes `` Message: "Domain `example.com` not found" ``
> 
> Transport Errors also double the error message. Giving the error message, then giving it again in parentheses
> 
> `Transport Error – https://api.sandbox.dnsimple.com/v2/whoami: Dns Failed: resolve dns name 'api.sandbox.dnsimple.com:443': failed to lookup address information: nodename nor servname provided, or not known(https://api.sandbox.dnsimple.com/v2/whoami: Dns Failed: resolve dns name 'api.sandbox.dnsimple.com:443': failed to lookup address information: nodename nor servname provided, or not known)`
> 
> I haven't tested all the possible errors yet but [I have started expanding the API examples to help](https://github.com/dnsimple/dnsimple-api-examples/pull/18) with integration tests.

I have cleaned the error messages. As it turns out when rust puts a string inside a string (from JSON) it adds quotation marks, which it then escapes. I remember googling this before and not finding any satisfactory solution (which was the reason for me leaving it as is in the first place). I've now made sure to parse the string coming from JSON and remove the `"` from it.

As for the `Transport` Error, there was a genuine bug in there where we only took one of the variables and used it to fill in both arguments; this has been fixed now as well.

File: src/dnsimple.rs [hunk 1]
```diff
@@ -501,24 +501,30 @@ impl Client {
...
     pub fn build_patch_request(&self, path: &&str) -> Request {
-        self._agent
+        let request = self
+            ._agent
             .request("PATCH", &self.url(path))
             .set("User-Agent", &self.user_agent)
-            .set("Accept", "application/json")
+            .set("Accept", "application/json");
+        self.add_headers_to_request(request)
     }
```
[the remark this answers is not in the document]
Comment (ecomba) [reply src/dnsimple.rs]: It is(n't). When I was cleaning the code for the Error messages I found some code duplication that was bugging me and I removed it

File: tests/zones_test.rs [hunk 2]
```diff
@@ -119,13 +119,10 @@ fn check_zone_distribution_error() {
...
-    let response = client
-        .zones()
-        .check_zone_distribution(account_id, zone)
-        .unwrap();
+    let response = client.zones().check_zone_distribution(account_id, zone);
 
     assert_eq!(
-        "Could not query zone, connection timed out",
-        response.errors.unwrap().message.unwrap()
+        "Message: Could not query zone, connection timed out",
```
Comment (onlyhavecans) [inline tests/zones_test.rs:125]: I think we shouldn't change the message and add `Message: `

**Review by onlyhavecans — CHANGES_REQUESTED** [review]
I have some small suggestions to make the dev experience better and more consistent. Also pushed minor changes for cleanup.

**Review by onlyhavecans — APPROVED** [review]
I made a few minor changes and I think this is good to go!
GG on this refactor! 👏🏻

**Pull request merged**